### PR TITLE
feat(cli): Android onboarding via Google OAuth (GCP + SA + Play invite automation)

### PR DESCRIPTION
**Pull request by WcaleNieWolny**
## Summary

Adds `capgo build init --platform android` — interactive TUI that handles the entire Android cloud-build setup for a developer in one flow.

Replaces the [manual ~30-step process documented at /docs/cli/cloud-build/android](https://capgo.app/docs/cli/cloud-build/android) (gcloud install → GCP project → enable API → service account → JSON key → Play Console invite → save credentials) with a single command. The output credential format (`PLAY_CONFIG_JSON` + keystore) is identical to the manual flow, so the existing fastlane-based build pipeline is untouched.

## What it does

1. **Keystore** — generates a PKCS#12 (`.p12`) via node-forge, or reuses an existing `.jks` / `.keystore` / `.p12` with auto-detected alias when possible
2. **Sign in with Google** — PKCE OAuth flow on a 127.0.0.1 loopback HTTP server. Scopes: `openid`, `userinfo.email`, `androidpublisher`, `cloud-platform` (Google has approved verification for Capgo's OAuth client)
3. **Play developer ID** — user pastes Play Console URL or numeric dev ID (Play Developer API has no listing endpoint)
4. **GCP project** — list user's projects or create a new `capgo-{slug}-{rand}` one
5. **Android package** — parses `android/app/build.gradle` (Groovy + Kotlin DSL) for `applicationId`s; falls back to manual input if Gradle is absent. Important because `CapacitorUpdater.appId` overrides often diverge from the Android applicationId
6. **Provisioning** — creates project (if new), enables `androidpublisher.googleapis.com`, creates `capgo-native-build` SA + JSON key, invites SA to Play Console with release-only permissions on the chosen package
7. **Save** — writes `ANDROID_KEYSTORE_FILE`, `KEYSTORE_*`, and `PLAY_CONFIG_JSON` to `~/.capgo-credentials/credentials.json` (same format as the manual flow)

## Depends on

- #2051 — `/private/config/builder` endpoint (already merged). The CLI fetches OAuth client credentials from there at runtime — **no secrets baked into the CLI binary**. Both `enabled: false` and the Capgo backend missing env vars are surfaced as a clear error pointing the user at the manual setup docs.

## Backend env vars expected

Already documented per #2051:

- \`GOOGLE_OAUTH_CLIENT_ID\` — Capgo's OAuth Desktop-app client ID
- \`GOOGLE_OAUTH_CLIENT_SECRET\` — that client's "secret" (treated as public per RFC 8252 §8.5; PKCE provides the actual security)

## Tests

29 new unit tests across 5 files in \`cli/test/test-android-*.mjs\` (keystore, OAuth, GCP, Play, Gradle parser). All offline — no network round trips.

## Test plan

- [ ] \`bun cli/test/test-android-keystore.mjs\`
- [ ] \`bun cli/test/test-android-oauth.mjs\`
- [ ] \`bun cli/test/test-android-gcp.mjs\`
- [ ] \`bun cli/test/test-android-play.mjs\`
- [ ] \`bun cli/test/test-android-gradle.mjs\`
- [ ] End-to-end: \`node cli/dist/index.js build init --platform android\` in a Capacitor project that already has a Play Console app entry
- [ ] Verify GCP project, service account, and SA key appear in console.cloud.google.com
- [ ] Verify SA email appears in Play Console → Users and permissions with release perms on the chosen package
- [ ] Verify resulting \`~/.capgo-credentials/credentials.json\` matches the manual-flow output and \`capgo build request --platform android\` succeeds with it

## Notes

- The OAuth client credentials previously committed to the now-archived \`Cap-go/CLI\` repo's \`feat/android-onboarding-oauth\` branch should be rotated in GCP Console — they were briefly in a public repo even though that repo is archived
- Tutorial video URL: https://www.youtube.com/watch?v=Y1_Ngj8hHLU (already linked in the dev-ID input step)
- OAuth verification approved by Google on 2026-05-02 for the four scopes above

## Summary by CodeRabbit

* **New Features**
  * Full Android onboarding TUI: step-driven, resumable progress, keystore generation/import with alias/password probing, Google OAuth (PKCE) sign-in, GCP project/service-account provisioning, and Play Console invite flow
  * Auto-detect Android app IDs from Gradle, macOS keystore file picker, masked password inputs, and optional build trigger after provisioning
  * CLI: interactive onboarding now supports ios or android via --platform and updated help text

* **Tests**
  * Added unit suites covering GCP, Gradle parsing, keystore, OAuth, and Play API helpers

[Review Change Stack](https://app.coderabbit.ai/change-stack/Cap-go/capgo/pull/2064)

### CODE REVIEW

File: cli/src/build/onboarding/android/ui/app.tsx [hunk 1]
```diff
@@ -0,0 +1,1535 @@
...
+                packageName: androidPackageChoice.packageName,
+                permissions: CAPGO_SA_APP_PERMISSIONS,
+              }],
+            })
+          }
+          catch (err) {
+            const msg = err instanceof Error ? err.message : String(err)
+            // Treat "already exists" style failures as success — the SA is
+            // already a user on this developer account from a prior run.
+            if (!/already|exists|duplicate/i.test(msg))
```
Comment (TateLyman) [inline cli/src/build/onboarding/android/ui/app.tsx:873]: This retry path treats any `already exists` invite response as success, but it never verifies that the existing Play Console user has the app-level grants for the package selected in this run. A common recovery case is: first run invites the service account with the wrong/missing package grant, the user reruns onboarding with the correct package, `users.create` returns an already-exists conflict, and this branch continues to save credentials even though the service account still cannot upload/releases for that package. The existing-user path needs to fetch/update the Play user grants (or fail with a clear manual-fix message) before marking the invite confirmed.

**Review by TateLyman — COMMENTED** [review]
One focused issue from the Android onboarding flow.

File: cli/src/build/onboarding/android/ui/app.tsx [hunk 2]
```diff
@@ -0,0 +1,1535 @@
...
+      {step === 'keystore-new-key-password' && (
+        <Box flexDirection="column" marginTop={1}>
+          <Text bold>Key password (press Enter to match store password):</Text>
+          <Newline />
+          <FilteredTextInput
+            placeholder="(hidden — same as store)"
+            filter=""
+            mask
+            onSubmit={(val) => {
+              const keyPw = val || keystoreStorePassword
```
Comment (TateLyman) [inline cli/src/build/onboarding/android/ui/app.tsx:1413]: For newly generated PKCS#12 keystores this lets the user choose a key password that differs from the store password, but `generateKeystore()` passes only `options.storePassword` to `forge.pkcs12.toPkcs12Asn1(...)` and never applies `options.keyPassword` to the private key. The credentials saved later therefore can contain `KEYSTORE_KEY_PASSWORD != KEYSTORE_STORE_PASSWORD` for a keystore that was actually written with only the store password, which will make the subsequent Android signing/build step fail for users who choose a custom key password. For generated `.p12` files, either force `keyPw = storePw`/hide this prompt, or teach the generator to really encode a separate key password and cover the mismatch in tests.

**Review by TateLyman — COMMENTED** [review]
One more generated-keystore edge case.

**Review by adamsardo — COMMENTED** [review]
One thing to tighten before merge: the Play invite fallback treats any `already/exists/duplicate` error from `inviteServiceAccount()` as success, but it never verifies that the existing Play Console user has the grant for the selected package.

The problematic path is in `cli/src/build/onboarding/android/ui/app.tsx` around the `inviteServiceAccount` call: if `users.create` returns an "already exists" style response, onboarding records `playInviteProvisioned`, continues to `saving-credentials`, and deletes the progress file. That is only safe if the existing service-account user already has the exact `androidPackageChoice.packageName` grant with the required app-level permissions.

A realistic failure case: a previous run invited `capgo-native-build@...` for package A, or with a narrower permission set. The user reruns onboarding for package B. Play returns that the user already exists, this code treats that as success, credentials are saved, and the later build/upload fails because the service account was never granted access to package B. The fix should either upsert/patch the user grants when the user already exists, or fetch the existing user/grants and require the selected package + permissions before marking the invite step complete.

**Review by mySebbe — COMMENTED** [review]
﻿One security isolation issue in the Android onboarding flow: the service-account identity is fixed per GCP project, not per app/package.

`DEFAULT_SERVICE_ACCOUNT_ID` is always `capgo-native-build`, and `ensureServiceAccount()` reuses an existing `capgo-native-build@<project>.iam.gserviceaccount.com`. The flow then creates a fresh JSON key for that same identity and saves it as this app's `PLAY_CONFIG_JSON`, while Play Console grants are added for the selected `androidPackageChoice.packageName`.

That means if the same GCP project is used to onboard multiple packages, the single service account accumulates grants for all of them. A JSON key saved/exported for app A can authenticate as the same service account and operate on package B after B is onboarded, so the per-app Capgo credential is not actually package-isolated.

This is separate from the existing "already exists" fallback issue: even when the invite/grant succeeds for each package, the saved keys still all belong to the same multi-package identity. Can we derive the service-account ID from the package/app, e.g. `capgo-${hash(packageName)}`, or otherwise create one service account per Play package and refuse/review reuse when the existing SA already has grants for other packages?

**Review by SpeedyArt — COMMENTED** [review]
I think this still grants an account-wide Play Console permission even though the flow is described as package-scoped release setup.

`inviteServiceAccount()` always receives `developerAccountPermissions: CAPGO_SA_DEVELOPER_PERMISSIONS`, and that constant includes `CAN_MANAGE_DRAFT_APPS_GLOBAL`. Google treats `developerAccountPermissions` as permissions for the whole developer account, and this value allows draft apps to be created/edited/deleted at the account level. The same request already includes per-package `grants` for the selected `androidPackageChoice.packageName`, including app-level `CAN_MANAGE_DRAFT_APPS`, so the global grant does not look necessary for the normal package-scoped onboarding path.

As written, the JSON key saved for one app can manage draft apps outside the selected package, even if the app-level grant is correctly scoped. That undercuts the "release-only permissions on the chosen package" model and is separate from the fixed service-account reuse issue already mentioned above. I’d remove `developerAccountPermissions` for the package-scoped flow if Play accepts users with app-level grants only, or gate the account-wide permission behind an explicit warning/confirmation. A regression around the invite body should assert that normal onboarding does not include `CAN_MANAGE_DRAFT_APPS_GLOBAL`.

File: cli/src/build/onboarding/android/play-api.ts [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+}): Promise<PlayInvitedUser> {
+  const body: Record<string, unknown> = {
+    email: args.serviceAccountEmail,
+  }
+  if (args.developerAccountPermissions?.length)
+    body.developerAccountPermissions = args.developerAccountPermissions
+  if (args.grants?.length) {
+    // Grant resource uses `appLevelPermissions`, NOT `permissions`.
+    // Ref: https://developers.google.com/android-publisher/api-ref/rest/v3/grants
+    body.grants = args.grants.map(g => ({
```
Comment (1412Atom) [inline cli/src/build/onboarding/android/play-api.ts:186]: The invite flow is sending the app grant inside `users.create`, but the Android Publisher `User` resource documents `grants[]` as output-only. That means onboarding can create/invite the service account while the package-level permissions are not actually created, so the later build/upload can still fail despite this step being marked confirmed.

The safer flow is to create/find the Play user first, then call the Grants API for `developers/{developerId}/users/{email}/grants/{packageName}` with `appLevelPermissions`. The `already exists` recovery path should also fetch/update that package grant before continuing, otherwise rerunning onboarding after a wrong/missing package grant will keep saving credentials that cannot upload releases.

Refs:
- https://developers.google.com/android-publisher/api-ref/rest/v3/users
- https://developers.google.com/android-publisher/api-ref/rest/v3/grants/create

**Review by 1412Atom — COMMENTED** [review]
I found one Play Console permission issue that can make the happy path look successful while leaving the service account without release access.

**Review by zinc-builds — APPROVED** [review]
## Security Review: #2064 — Android OAuth onboarding

### Looks Good ✅
- **PKCE S256 flow** for OAuth token exchange — industry standard, prevents authorization code interception
- **Loopback server on 127.0.0.1 only** — `LOOPBACK_HOST` hardcoded, no external network exposure
- **crypto.randomBytes** for keystore/store/key passwords — cryptographically strong
- **encodeURIComponent** on all user-supplied values in GCP/Play API URLs — prevents path injection
- **JSON response validation** after every GCP/Play API call — catches unexpected responses before they propagate
- **Service account key base64** never written to permanent credentials store (marked with underscore prefix convention)
- **Progress file permissions 0o600** via `writeFileAtomic` — not world-readable
- **ensureSecureDirectory with 0o700** for onboarding dir — prevents other users from reading progress files
- **Gradle parser reads local files only** — no remote fetching, no command injection
- **sanitizeGcpProjectDisplayName** strips disallowed characters before sending to GCP API

### Warnings ⚠️
- **Sensitive tokens in progress file**: `_oauthRefreshToken`, `_serviceAccountKeyBase64`, `_keystoreBase64` persist on disk at `~/.capgo-credentials/onboarding/android-{appId}.json` (0o600) until onboarding completes. If the process crashes or the user Ctrl-Cs mid-flow, these tokens remain on disk. The `deleteAndroidProgress` call at the end of successful onboarding does clean up, but crash recovery does not guarantee it.
- **Math.random for project ID suffix** (generateProjectId): Used for collision avoidance only (not security), and the comment acknowledges this. Acceptable for this use case, but consider `crypto.randomInt` for consistency with the rest of the codebase crypto hygiene.
- **Refresh token in memory**: `refreshTokenState` held in React component state for the full onboarding session. If a future code change logs component state for debugging, this token could leak.

### Suggestions 💡
- Consider clearing sensitive underscore-prefixed fields from progress immediately after use (e.g., clear `_oauthRefreshToken` after SA key is provisioned, clear `_keystoreBase64` after credentials are saved to permanent store)
- Add a signal handler (SIGINT/SIGTERM) that cleans up the progress file before exit
- The `open` package is imported but the fallback path in `runOAuthFlow` already handles browser-open failures gracefully — consider adding a timeout for the browser-open call to avoid hanging on headless systems

### Summary
Well-structured, security-conscious code. The PKCE flow, loopback binding, and file permission practices are solid. The primary risk is token persistence in the progress file across crashes, which is partially mitigated by 0o600 permissions and the stated intent to wipe on completion. No auth bypass, no injection, no data exposure to external parties. **Approved.**

**Comment by WcaleNieWolny** [comment]
What the actual fuck.... this is a draft PR...

File: cli/src/build/onboarding/android/gcp-api.ts [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    keyAlgorithm?: string
+  }>({
+    method: 'POST',
+    url: `${IAM_ENDPOINT}/projects/${encodeURIComponent(args.projectId)}/serviceAccounts/${encodeURIComponent(args.serviceAccountEmail)}/keys`,
+    accessToken: args.accessToken,
+    body: { privateKeyType: 'TYPE_GOOGLE_CREDENTIALS_FILE', keyAlgorithm: 'KEY_ALG_RSA_2048' },
+  })
+  if (!res.privateKeyData)
+    throw new Error('Google did not return privateKeyData on the SA key — refusing to continue without the key file')
+  return { name: res.name, privateKeyDataBase64: res.privateKeyData }
```
[the remark this answers is not in the document]
Comment (WcaleNieWolny) [reply cli/src/build/onboarding/android/gcp-api.ts:321]: Not possible to fix

File: cli/src/build/onboarding/android/ui/app.tsx [hunk 5]
```diff
@@ -0,0 +1,1798 @@
...
+            uniqueId: sa.uniqueId,
+          }
+          setServiceAccountProvisioned(saProv)
+          addSetupStatus(saCreated ? `✔ Service account created — ${sa.email}` : `✔ Service account exists — ${sa.email}`)
+
+          // Step D: create a fresh JSON key for the SA
+          addSetupStatus('Creating service-account JSON key...')
+          const key = await createServiceAccountKey({
+            accessToken: tok,
+            projectId: projectChoice.projectId,
```
[the remark this answers is not in the document]
Comment (WcaleNieWolny) [reply cli/src/build/onboarding/android/ui/app.tsx:838]: Not possible to fix

File: cli/src/build/onboarding/command.ts [hunk 6]
```diff
@@ -38,13 +93,24 @@ export async function onboardingBuilderCommand(options: OnboardingBuilderOptions
...
 
-  // Launch Ink app
+  if (platform === 'android') {
+    const androidProgress = await loadAndroidProgress(appId)
+    const { waitUntilExit } = render(
+      React.createElement(AndroidOnboardingApp, {
+        appId,
+        initialProgress: androidProgress,
+        androidDir,
+      }),
```
[the remark this answers is not in the document]
Comment (WcaleNieWolny) [reply cli/src/build/onboarding/command.ts:106]: Fixed
